es5-shim.js

```diff
@@ -175,11 +177,160 @@ var ES = {
     }
 };
 
+// Check failure of by-index access of string characters (IE < 9)
+// and failure of `0 in boxedString` (Rhino)
+var boxedString = $Object('a');
+var splitString = boxedString[0] !== 'a' || !(0 in boxedString);
+
 //
 // Function
 // ========
 //
 
+// Tests for inconsistent or buggy `[[Class]]` strings.
+/* eslint-disable no-useless-call */
+var hasToStringTagBasicBug = to_string.call() !== '[object Undefined]' || to_string.call(null) !== '[object Null]';
+/* eslint-enable no-useless-call */
+var hasToStringTagLegacyArguments = to_string.call(arguments) !== '[object Arguments]';
+var hasToStringTagInconsistency = hasToStringTagBasicBug || hasToStringTagLegacyArguments;
+// Others that could be fixed:
+// Older ES3 native functions like `alert` return `[object Object]`.
+// Inconsistent `[[Class]]` strings for `window` or `global`.
+
+var hasApplyArrayLikeDeficiency = (function () {
+    var arrayLike = { length: 4, 0: 1, 2: 4, 3: true };
+    var expectedArray = [1, undefined, 4, true];
+    var actualArray;
+    try {
+        actualArray = (function () {
+            // `array_slice` is safe to use here, no known issue at present.
+            return array_slice.apply(arguments);
+        }.apply(null, arrayLike));
+    } catch (e) {
+        if (to_string.call(actualArray) !== '[object Array]' || actualArray.length !== arrayLike.length) {
+            return true;
+        }
+        while (expectedArray.length) {
+            if (actualArray.pop() !== expectedArray.pop()) {
+                return true;
+            }
+        }
+    }
+    return false;
+}());
+
+var shouldPatchCallApply = hasToStringTagInconsistency || hasApplyArrayLikeDeficiency;
+
+if (shouldPatchCallApply) {
+    // Constant. ES3 maximum array length.
+    var MAX_ARRAY_LENGTH = 4294967295;
```

Comment:
`Math.pow(2, 32) - 1`. This is the maximum value that any ES3 environment that I have tested supports (I don't know if it is in the spec somewhere). Then in ES5 it changes to `Number.MAX_SAFE_INTEGER` for arrays and strings (this is in the spec as far I remember). Arrays seem to follow this but strings still seem to be limited to 65k characters, it is a similar story with arguments, I don't think any environment supports more than 65k args (and many support much less). So this seemed a sensible number to choose to check if it was array-like, the smallest max array length? We could set it to `Number.MAX_SAFE_INTEGER`, but this patch is only going to be applied on ES3 environments, I don't know, I am in two minds about what it should be?